text/0000-pub-macro-rules.md

```diff
@@ -0,0 +1,300 @@
+    * > Thus, code that compiles without warnings on the previous edition (under the latest compiler release) will compile without errors on the next edition (modulo the usual caveats about type inference changes and so on).
+    * This needs to be weighed against the disadvantages of the multi-step transition plan discussed in the alternatives section.
+* Many have expressed informal desire for a less "exciting" edition than Rust 2018. While transitioning in one edition may avoid a less than ideal temporary state before the full transition is achieved, it is also less of a big change all at once. 
+
+# Rationale and alternatives
+
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+No alternatives are currently being explored for the end vision of this RFC. The alternatives mentioned here are alternatives to the transition which all will ultimately end in fully path scoped macros.
+
+## Deprecating `#[macro_export]` and `#[macro_rules]` in Rust 2021
+
+An alternative to the full transition to path scoping is to deprecate the use of `#[macro_export]` and `#[macro_use]` in favor of path scoping. 
+
+In particular this means the following:
+* By default, emit a warning when using `#[macro_export]` or `#[macro_use]` suggesting the user to use `pub` annotations and access macros by path instead.
+* Any macro marked with `pub` stops following textual scoping rules and can be accessed by path like any other item. 
+    * One consequence of this means annotating a module with `#[macro_use]` will not have any impact on the visibility of macros marked as `pub` inside that module.
+    * We can potentially warn against the mixed use of `pub` and `#[macro_use]`
+* Any macro not marked with `pub` continues to follow textual scoping rules
+    * This includes shadowing and not being usable before being defined
+* Marking a macro as `pub` and annotating with `#[macro_export]` is a hard error.
+
+### Advantages 
+
+This has the advantage not forcing users to upgrade their code at the point of moving to a new edition. They will receive warnings but their code continues to compile and can be gradually transitioned. 
+
+Thus this avoids the [drawbacks](#transition-drawbacks) of the proposed full transition above.
+
+### Disadvantages
+
+There are several downsides to this proposal:
+* The existence of two scoping systems at once can be confusing especially if users mix the two usages. 
+    * Currently, it is possible to "convert" a macro to use path based scoping [by reexporting a macro from a module](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=275c52e038b5ab3db526a605874fbd17), but this use case is rare. The proposal above would likely see much deeper mixing of the two systems. 
+* Even if the user never uses the deprecated annotations, macros defined and used locally to a module still follow textual scoping rules unlike every other item in the language. 
```

Comment:
This approach is definitely very interesting. One hiccup to keep in mind, though, is the following:

```rust
fn foo () {}
mod scoped {
    use super::foo;

    /* textual */ macro_rules! foo {() => ()}
}
/* textual */ macro_rules! foo {() => ()}
```

This is code, that, IIUC, would not trigger warnings on this hypothetical 2021 edition, but then the next edition that removes textual scoping would have to deal with it, where it features the shadowing problem (which can be fixed as mentioned above in the RFC, by renaming the macro defined inside `scoped` to, say, `foo2`, and so on).